marker_api/src/ast/common/span.rs

```diff
@@ -2,133 +2,469 @@ use std::marker::PhantomData;
+    /// should be filled out. For example, if you want to snip an expression, you
+    /// should use `<expr>` as the default value.
+    ///
+    /// If you're planning to use this snippet in a suggestion, consider using
+    /// [`snippet_with_applicability`](Self::snippet_with_applicability) instead.
+    pub fn snippet_or(&self, default: &str) -> String {
+        self.snippet().unwrap_or(default).to_string()
+    }
+
+    /// Adjusts the given [`Applicability`] according to the context and returns the
+    /// code snippet that this [`Span`] refers to or the given default if the
+    /// snippet is unavailable.
+    ///
+    /// For the placeholder, it's recommended to use angle brackets with information
+    /// should be filled out. A placeholder for an expression should look like
+    /// this: `<expr>`
+    ///
+    /// The applicability will never be upgraded by this method. When you draft
+    /// suggestions, you'll generally start with the highest [`Applicability`]
+    /// your suggestion should have, and then use it with this snippet function
+    /// to adjust it accordingly. The applicability is then used to submit the
+    /// suggestion to the driver.
+    pub fn snippet_with_applicability(&self, placeholder: &str, applicability: &mut Applicability) -> String {
+        if *applicability != Applicability::Unspecified && self.is_from_expansion() {
+            *applicability = Applicability::MaybeIncorrect;
+        }
+        self.snippet()
+            .unwrap_or_else(|| {
+                if matches!(
+                    *applicability,
+                    Applicability::MachineApplicable | Applicability::MaybeIncorrect
+                ) {
+                    *applicability = Applicability::HasPlaceholders;
+                }
+                placeholder
+            })
+            .to_string()
     }
 
-    pub fn is_from_macro(&self) -> bool {
-        matches!(self.source, SpanSource::Macro(..))
+    /// Returns the length of the this [`Span`] in bytes.
+    pub fn len(&self) -> usize {
+        (self.start.0 - self.end.0)
+            .try_into()
+            .expect("Marker is not compiled for usize::BITs < 32")
     }
 
     /// Returns `true` if the span has a length of 0. This means that no bytes are
     /// inside the span.
     pub fn is_empty(&self) -> bool {
-        self.start == self.end
+        self.len() == 0
     }
 
-    /// Returns true, if both spans originate from the same source. For example, this can be the
-    /// same source file or macro expansion.
-    pub fn is_same_source(&self, other: &Span<'ast>) -> bool {
-        self.source == other.source
-    }
-
-    pub fn start(&self) -> usize {
+    /// Returns the start position of this [`Span`].
+    pub fn start(&self) -> SpanPos {
         self.start
     }
 
-    pub fn set_start(&mut self, start: usize) {
+    /// Sets the start position of this [`Span`].
+    pub fn set_start(&mut self, start: SpanPos) {
```

Comment:
Having functions to nicely combine spans is on the todo list I wanted to create a good-first-issue for it, once this PR is merged. Marker currently has very few issues, this can be picked up easily, that's why I haven't added them yet.

For the dangerous part, yeah, in Clippy we had a few panics due to that, but sometimes this is just needed to create better suggestions. I would rather not lock users out of creating great suggestions, especially if Marker might be used for automatic migrations. Marker can easily have a lint, to recommend users to use another method instead